tests/models/litter/test_litter_model.py

```diff
@@ -77,180 +48,139 @@ def test_litter_model_initialization(caplog, dummy_litter_data):
-        (
-            DEBUG,
-            "litter model: required var 'lignin_woody' checked",
-        ),
-        (
-            DEBUG,
-            "litter model: required var 'lignin_below_structural' checked",
-        ),
-        (
-            ERROR,
-            "Negative pool sizes found in: ",
-        ),
-    )
+    from virtual_rainforest.models.litter.constants import LitterConsts
+    from virtual_rainforest.models.litter.litter_model import LitterModel
 
     with pytest.raises(InitialisationError):
-        # Make four cell grid
-        litter_data = deepcopy(dummy_litter_data)
         # Put incorrect data in for lmwc
-        litter_data["litter_pool_above_metabolic"] = DataArray(
+        dummy_litter_data["litter_pool_above_metabolic"] = DataArray(
             [0.05, 0.02, -0.1], dims=["cell_id"]
         )
 
-        LitterModel(litter_data, pint.Quantity("1 week"), 2, 10, constants=LitterConsts)
+        LitterModel(
+            data=dummy_litter_data,
+            update_interval=pint.Quantity("1 week"),
+            soil_layers=2,
+            canopy_layers=10,
+            constants=LitterConsts,
+        )
 
-    # Final check that expected logging entries are produced
-    log_check(caplog, expected_log_entries)
+    # Final check that the last log entry is as expected
+    log_check(
+        caplog,
+        expected_log=((ERROR, "Negative pool sizes found in: "),),
+        subset=slice(-1, None, None),
+    )
 
 
 def test_litter_model_initialization_bad_lignin_bounds(caplog, dummy_litter_data):
     """Test `LitterModel` initialization fails for lignin proportions not in bounds."""
 
-    expected_log_entries = (
-        (
-            INFO,
-            "Replacing data array for 'lignin_woody'",
-        ),
-        (
-            DEBUG,
-            "litter model: required var 'litter_pool_above_metabolic' checked",
-        ),
-        (
-            DEBUG,
-            "litter model: required var 'litter_pool_above_structural' checked",
-        ),
-        (
-            DEBUG,
-            "litter model: required var 'litter_pool_woody' checked",
-        ),
-        (
-            DEBUG,
-            "litter model: required var 'litter_pool_below_metabolic' checked",
-        ),
-        (
-            DEBUG,
-            "litter model: required var 'litter_pool_below_structural' checked",
-        ),
-        (
-            DEBUG,
-            "litter model: required var 'lignin_above_structural' checked",
-        ),
-        (
-            DEBUG,
-            "litter model: required var 'lignin_woody' checked",
-        ),
-        (
-            DEBUG,
-            "litter model: required var 'lignin_below_structural' checked",
-        ),
-        (
-            ERROR,
-            "Lignin proportions not between 0 and 1 found in: ",
-        ),
-    )
+    from virtual_rainforest.models.litter.constants import LitterConsts
+    from virtual_rainforest.models.litter.litter_model import LitterModel
```

Comment:
If I understood @dalonsoa correctly, I think the idea is that as much as possible of the environment used by each test should be isolated within it? With some `vr` things - notably custom `Exceptions` - we have to import them at the module level so they can be used in parameterisation, but otherwise tests should be self contained? Is that right?

Reply:
That's correct. As much as possible, setting up the tests (as oppose to running them) should not depend on the package under test itself, and that means having all the imports - even if repeated - within the tests. This avoids side effects (keeping appropriate isolation) as well as preventing the whole test suit to fail at setup time just because a failure of the package under test. 

Having said that, as David points out, there are specific use cases where that is not possible, but it should be the general approach.